### DISCUSSION TITLE
Necessity of scaling gradient of variables registered local

### DISCUSSION THREAD

**Discussion by skyw** [horovod/horovod#3705] [General]
Continue the discussion in https://github.com/horovod/horovod/pull/3695.

In the most common way of model parallel embedding, i.e. different embedding tables on different GPU/CPU, with a MLP sitting on top. An alltoall is between embedding and MLP to switch from/to model and data parallel. That creates a trap in the loss scale. gradient of MLP in data parallel fashion is divided by world size (or all-reduce domain size), but gradient of embedding in model parallel region are not. 

Below is a simple reproduce, the same model trained with the same data get different results just with different parallelism.
```python
import numpy as np
import tensorflow as tf
from tensorflow import keras

import horovod.tensorflow as hvd

class DummyDPModel(keras.Model):
  def __init__(self):
    super().__init__()

    self.embeddings = [
        keras.layers.Embedding(7, 3, embeddings_initializer=keras.initializers.Constant(1.)),
        keras.layers.Embedding(9, 3, embeddings_initializer=keras.initializers.Constant(2.))]
    self.concat = keras.layers.Concatenate(1)
    self.dense = keras.layers.Dense(
        1, use_bias=False, kernel_initializer=keras.initializers.Constant(1.))

  def call(self, inputs):
    x = self.concat([embedding(ids) for embedding, ids in zip(self.embeddings, inputs)])
    out = self.dense(x)
    return out

class DummyMPModel2Devices(keras.Model):
  def __init__(self):
    # For demonstation purpose, only supports 2 way parallel
    assert hvd.size() == 2
    super().__init__()

    if hvd.rank() == 0:
      self.embedding = keras.layers.Embedding(7, 3, embeddings_initializer=keras.initializers.Constant(1.))
    else:
      self.embedding = keras.layers.Embedding(9, 3, embeddings_initializer=keras.initializers.Constant(2.))

    self.concat = keras.layers.Concatenate()
    self.dense = keras.layers.Dense(
        1, use_bias=False, kernel_initializer=keras.initializers.Constant(1.))

  def call(self, inputs):
    x = self.embedding(inputs)
    y = hvd.alltoall(x)
    dp_batch_size = inputs.shape[0] // hvd.size()
    out = self.dense(tf.reshape(y, [dp_batch_size, -1]))
    return out


def main():
  hvd.init()
  hvd_rank = hvd.rank()
  hvd_size = hvd.size()
  gpus = tf.config.experimental.list_physical_devices('GPU')
  for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)
  if gpus:
    tf.config.experimental.set_visible_devices(gpus[hvd.local_rank()], 'GPU')

  dp_model = DummyDPModel()
  mp_model = DummyMPModel2Devices()

  optimizer = tf.keras.optimizers.SGD(learning_rate=1)
  bce = keras.losses.BinaryCrossentropy(reduction=keras.losses.Reduction.NONE, from_logits=True)
  labels = tf.constant([1.])

  if hvd_rank == 0:
    dp_inputs = [
        tf.constant([hvd_rank + 1], dtype=tf.int64),
        tf.constant([hvd_rank], dtype=tf.int64)
    ]
  else:
    dp_inputs = [
        tf.constant([hvd_rank + 2], dtype=tf.int64),
        tf.constant([hvd_rank + 3], dtype=tf.int64)
    ]
  dp_out = dp_model(dp_inputs)

  # "Transpose" input from data parallel to model parallel
  mp_inputs = hvd.alltoall(dp_inputs)
  mp_out = mp_model(mp_inputs)

  # Model and data parallel should generate the same results
  tf.debugging.assert_equal(mp_out, dp_out)

  @tf.function
  def dp_train_step(inputs):
    with tf.GradientTape() as tape:
      predictions = dp_model(inputs)
      loss = tf.math.reduce_mean(bce(labels, predictions))
    tape = hvd.DistributedGradientTape(tape)
    gradients = tape.gradient(loss, dp_model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, dp_model.trainable_variables))
    return loss

  dp_loss = dp_train_step(dp_inputs)

  @tf.function
  def mp_train_step(inputs):
    with tf.GradientTape() as tape:
      predictions = mp_model(inputs)
      loss = tf.math.reduce_mean(bce(labels, predictions))
    tape = hvd.DistributedGradientTape(tape)
    tape.register_local_source(mp_model.embedding.weights[0])
    gradients = tape.gradient(loss, mp_model.trainable_variables)
    optimizer.apply_gradients(zip(gradients, mp_model.trainable_variables))
    return loss

  mp_loss = mp_train_step(mp_inputs)

  # loss should be the same
  tf.debugging.assert_equal(mp_out, dp_out)

  # MLP weights should also be the same
  tf.debugging.assert_equal(dp_model.dense.weights, mp_model.dense.weights)

  # But embedding weights are different. Gradients were off by world size
  if hvd_rank == 0:
    print(dp_model.embeddings[0].weights[0])
    print(mp_model.embedding.weights[0])
    tf.debugging.assert_near(dp_model.embeddings[0].weights[0], mp_model.embedding.weights[0], atol=0, rtol=1e-5)

  if hvd_rank == 1:
    print(dp_model.embeddings[1].weights[0])
    print(mp_model.embedding.weights[0])
    tf.debugging.assert_near(dp_model.embeddings[1].weights[0], mp_model.embedding.weights[0], atol=0, rtol=1e-5)

if __name__ == '__main__':
  main()
```

It is not technically wrong, but may not be what is expected. Either scaling gradient of embedding, or adjusting learning rate can make them align. But I think an option should be provided to make adjustment easy, especially for DistributedOptimizer with register_local_var capability, whose design point is hiding things behind without needing users to worry about them. Then model.fit can just work. In which case, either scaling gradient or adjusting learning rate seems to defeat the goal.

@MrAta @maxhgerlach @romerojosh

**Comment by MrAta** [comment]
@skyw thanks for taking the time to write the reproducible code. 
I commented out the assertions in your code to inspect the weights for embedding tables manually, and here is what I get:
```bash
[1,0]<stdout>:@@@@@@@@@@@@@@@@@@@@DP EM0 WEIGHTS @@@@@@@@@@@@@@@@@@@@@@
[1,0]<stdout>:<tf.Variable 'dummy_dp_model/embedding/embeddings:0' shape=(7, 3) dtype=float32, numpy=
[1,0]<stdout>:array([[1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.0118564, 1.0118564, 1.0118564],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.0118564, 1.0118564, 1.0118564],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.       , 1.       , 1.       ]], dtype=float32)>
[1,0]<stdout>:@@@@@@@@@@@@@@@@@@@@MP EM0 WEIGHTS @@@@@@@@@@@@@@@@@@@@@@
[1,0]<stdout>:<tf.Variable 'dummy_mp_model2_devices/embedding_2/embeddings:0' shape=(7, 3) dtype=float32, numpy=
[1,0]<stdout>:array([[1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.0237129, 1.0237129, 1.0237129],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.0237129, 1.0237129, 1.0237129],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.       , 1.       , 1.       ],
[1,0]<stdout>:       [1.       , 1.       , 1.       ]], dtype=float32)>

[1,1]<stdout>:@@@@@@@@@@@@@@@@@@@@DP EM1 WEIGHTS  @@@@@@@@@@@@@@@@@@@@@@
[1,1]<stdout>:<tf.Variable 'dummy_dp_model/embedding_1/embeddings:0' shape=(9, 3) dtype=float32, numpy=
[1,1]<stdout>:array([[2.0006182, 2.0006182, 2.0006182],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.0006182, 2.0006182, 2.0006182],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ]], dtype=float32)>
[1,1]<stdout>:@@@@@@@@@@@@@@@@@@@@MP EM  WEIGHTS @@@@@@@@@@@@@@@@@@@@@@
[1,1]<stdout>:<tf.Variable 'dummy_mp_model2_devices/embedding_2/embeddings:0' shape=(9, 3) dtype=float32, numpy=
[1,1]<stdout>:array([[2.0012364, 2.0012364, 2.0012364],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.0012364, 2.0012364, 2.0012364],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ],
[1,1]<stdout>:       [2.       , 2.       , 2.       ]], dtype=float32)>
```
As it can be seen in the output, the weights are off by `1e-3` not `world size`. Am I missing anything here?

**Comment by skyw** [comment]
Hi, @MrAta 
It was gradients that were off by world size, not weight itself. In the printed out values, weights were initialized to 1 and 2, if you compare the updated amount
(1.0237129 - 1) / (1.0118564 -1) = 2
(2.0012364 - 2) / ( 2.0006182 - 2) = 2

Print gradient before weight update will show the same.

**Comment by plliao** [comment]
Hi @skyw,

Thanks for the discussion. I have one observation. The data parallelism model looks a little bit weird o me. 
It does not concat the output from two embedding tables. Instead, it reads data from embedding 1 as the first sample and embedding 2 as the second sample. If you print out the shape of the dense layers, it only has shape of 3 (not 6).

Therefore, I guess what happened is that.

For the data parallelism example, the gradient was first averaged by local batch size in bce loss function and then averaged across ranks. It was divided by 4 in total. The correct divisor should be 2 as the global batch size is 2 in this case.

For the model parallelism example, the gradient was only averaged by local batch size. In this case, the local batch size to embedding equals global batch size. Hence, we have the correct divisor for model parallelism example.

Let me know if I misunderstand anything. Thanks!

**Reply by skyw** [reply]
You are right, I concatenated along the wrong axis. I made an edit/correction to the posted code.

Gradient (or loss) is averaged by global batch size (2) in data parallel case. while in the model (rather hybrid) parallel case, gradient of embedding is averaged by only data parallel batch size (1), while gradient of MLP is averaged by global batch size (2) so they are inconsistent.

**Reply by yundai424** [reply]
Imagine the per-GPU batch size is `k` and the world size (i.e. # GPUs) is `p`. My understanding is that, in DP case, the per-rank loss out of bce and `reduce_mean` is divided by `k`, so are the per-rank MLP gradients as well as per-rank embedding gradients. Afterwards when horovod is doing `allreduce` on MLP gradients and embedding gradients, both of them will be summed and further divided by `p`. The way horovod does `Average` is to modify the `postscale_factor` ([ref](https://github.com/horovod/horovod/blob/master/horovod/common/ops/nccl_operations.cc#L202)) which happens **after** the per-rank gradient is computed, so it doesn't seem to be the case that we're dividing it by 4 in total.

In MP scenario, the loss out of BCE and `reduce_mean` is still divided by `k`, and so is the per-rank MLP gradient. Going further down the backprop phase, we `alltoall` the MLP gradients, and then "reformat" them inside of embedding lookup gradient op (some slicing, concat etc. operations) to get the embedding gradients w.r.t. per-rank features on **all `p*k` examples**. However we don't have any further division by `p` here, thus the discrepancy. 🤔

**Reply by plliao** [reply]
Ok, I am wrong. Actually, the global batch size is 4 for the wrong DP example. Therefore, dividing the gradient by 4 is correct.
I miss the point that the gradient of the embedding is divided by the prediction batch size (**k**), instead of the embedding output batch size (**p * k**). I agree we should scale the gradient of local variable by the world size (**p**).

**Comment by MrAta** [comment]
I'll raise a follow up PR to fix this in #3695 and #3643. Thanks @skyw for bringing it up.

**Reply by skyw** [reply]
Thanks.
Maybe also add the option to gradient tape? which also has an register local capability.

**Reply by MrAta** [reply]
Yes, both tape and optimizer.
